File: solutions/4285-smallest-stable-index-ii/smallest-stable-index-ii.js
var firstStableIndex = function (nums, k) {
    const n = nums.length;
    const minValue = new Array(n);
    minValue[n - 1] = nums[n - 1];
    for (let i = n - 2; i >= 0; i--) {
        minValue[i] = Math.min(minValue[i + 1], nums[i]);
    }

    let maxValue = 0;
    for (let i = 0; i < n; i++) {
        maxValue = Math.max(maxValue, nums[i]);
        if (maxValue - minValue[i] <= k) {
            return i;
        }
    }
    return -1;
};

File: solutions/4285-smallest-stable-index-ii/smallest-stable-index-ii.rs
impl Solution {
    pub fn first_stable_index(nums: Vec<i32>, k: i32) -> i32 {
        let n = nums.len();
        let mut min_value = vec![0; n];
        min_value[n - 1] = nums[n - 1];
        for i in (0..n - 1).rev() {
            min_value[i] = min_value[i + 1].min(nums[i]);
        }

        let mut max_value = 0;
        for i in 0..n {
            max_value = max_value.max(nums[i]);
            if max_value - min_value[i] <= k {
                return i as i32;
            }
        }
        -1
    }
}